[numpy] Add torch.nan_to_num

kshitij12345 opened the pull request:
Reference #42515

TODO:
* [x] Add tests
* [x] Add docs

kshitij12345 committed:
add nan_to_num kernel impl

kshitij12345 committed:
Merge branch 'master' into develop/numpy/nan_to_num

kshitij12345 committed:
extra machinery for c10::Half support

kshitij12345 committed:
add docs

kshitij12345 committed:
update overrides.py

kshitij12345 committed:
update signature

kshitij12345 committed:
support gradient

kshitij12345 committed:
add test vs numpy

kshitij12345 committed:
rename pos_inf -> posinf, neg_inf -> neginf

kshitij12345 committed:
Merge branch 'master' into develop/numpy/nan_to_num

kshitij12345 committed:
remove merge stray

kshitij12345 commented:
@mruberry Please take a look :)

Have made a independent test as I dont think that `UnaryUfuncInfo` can take care of `nan`, `posinfo`, `neginfo` arguments.

`gradcheck` works but for `gradgradcheck`, we get 
```
RuntimeError: isposinf(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

Should we just manually add `zeros_like` for the `nan_to_num_backward`?

albanD commented:
Hi,

For the formula issue, it is an oversight on our side sorry.
Could you add to [this list](https://github.com/pytorch/pytorch/blob/a6895d43b6e59b66484f6fab967962c5db60ce01/tools/autograd/gen_variable_type.py#L142) a new line stating that functions returning boolean type are not differentiable and add the strings "isnan", "isposinf", "isneginf" and "isinf" there. That will solve your issue.

Also for the backward function, I think you can use `isinf` instead of doing two calls for the pos and neg infs.

kshitij12345 committed:
update argument names

kshitij12345 committed:
address comment

* update derivative not required ops.
* add gradgradcheck to test.
* use isinf.

kshitij12345 commented:
@albanD 
> Could you add to this list a new line stating that functions returning boolean type are not differentiable and add the strings "isnan", "isposinf", "isneginf" and "isinf" there. That will solve your issue.

Thanks that worked!

> Also for the backward function, I think you can use isinf instead of doing two calls for the pos and neg infs.

Ah. Right! Thanks!

albanD commented on test/test_autograd.py:
```diff
@@ -4527,6 +4527,33 @@ def test(inp, inp_dtype, out_dtype):
+    def test_nan_to_num(self):
+        a = torch.randn(3, 3, 3, 3)
+        with torch.no_grad():
+            a[torch.randn_like(a) < 0.2] = float('nan')
```
If you use randn here, <0.2 will be most of the Tensor no? Maybe you want something smaller?

kshitij12345 replied:
Right. Was supposed to be `torch.rand_like(a) < 0.2`

albanD reviewed (commented):
The autograd part looks good to me. Just small comment on the test.

I'll let @mruberry review the rest.

kshitij12345 committed:
replace randn < 0.2 -> rand < 0.2

mruberry commented on aten/src/ATen/NumericUtils.h:
```diff
@@ -50,6 +50,39 @@ inline C10_HOST_DEVICE bool _isnan(at::BFloat16 val) {
+
+template <typename T,
+          typename std::enable_if<c10::is_complex<T>::value, int>::type = 0>
+inline bool _isfinite(T val) {
```
The other functions here are annotated with C10_HOST_DEVICE - why isn't this one?

kshitij12345 replied:
Oops. Missed it. Thanks!

mruberry commented on aten/src/ATen/NumericUtils.h:
```diff
@@ -50,6 +50,39 @@ inline C10_HOST_DEVICE bool _isnan(at::BFloat16 val) {
+
+
+inline C10_HOST_DEVICE bool _isfinite(at::BFloat16 val) {
+  return at::_isfinite(float(val));
```
Should this be implemented as a conversion -> isfinite call or using std::numeric_limits, which PyTorch implements for bfloat16 (and float16 above, I think). 

If it's going to use a cast it should be a C++ style cast: `static_cast<float>(...)`

kshitij12345 replied:
I think using the `isfinite` is more consistent. Will use static_cast.

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -373,6 +373,34 @@ Tensor& logit_(Tensor& self, c10::optional<double> eps) {
+Tensor& nan_to_num_out(
+    Tensor& result,
+    const Tensor& self,
+    c10::optional<double> nan,
```
Should this be named 'num' (since it's "nan to num")?

mruberry replied:
Ug... but you're correct because in NumPy it's named "nan." Not a great name but you're right we should be consistent.

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -371,6 +371,37 @@ static void polygamma_kernel(TensorIterator& iter, int64_t n) {
+    c10::optional<double> nan,
+    c10::optional<double> pos_inf,
+    c10::optional<double> neg_inf) {
+  if (c10::isIntegralType(iter.dtype(), true)) {
```
Should nan_to_num just call clone on non-float, non-complex types?

kshitij12345 replied:
I was thinking about it. But since `out` is a kwarg, for correctly handling and using the passed tensor, I felt using the iterator and identity operator seemed better.

mruberry replied:
Hm... wouldn't it just be this check with:

```
result.copy_(self);
```

? 

That seems pretty straightforward.

mruberry replied:
(you might have to resize result using `resize_output()` first)

mruberry commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -186,6 +187,36 @@ void clamp_max_kernel_cuda(TensorIterator& iter, Scalar max_value) {
+    c10::optional<double> nan,
+    c10::optional<double> pos_inf,
+    c10::optional<double> neg_inf) {
+  if (c10::isIntegralType(iter.dtype(), true)) {
```
Same question about integer dispatch here.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,45 @@ def merge_dicts(*dicts):
+           r"""
+nan_to_num(input, nan=0.0, posinf=None, neginf=None, *, out=None) -> Tensor
+
+Replace NaN with zero and infinity with large finite numbers (default behaviour)
```
What about this to replace both paragraphs? 

"Replaces \`\`NaN`\`\, positive infinity, and negative infinity values in :attr:`input` with the values specified by :attr:\`nan\`, :attr:\`posinf\`, and :attr:\`neginf\`, respectively. By default, \`\`NaN\`\`s are replaced with zero, positive infinity is replaced with the greatest finite value representable by :attr:\`input\`'s dtype, and negative infinity is replaced with the least finite value representable by :attr:\`input\`'s dtype."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,45 @@ def merge_dicts(*dicts):
+
+Args:
+    {input}
+    nan (float or optional): Value to be used to fill NaN values.
```
nan (Number): the value to replace \`\`NaN\`\`s with.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,45 @@ def merge_dicts(*dicts):
+    {input}
+    nan (float or optional): Value to be used to fill NaN values.
+    If no value is passed then NaN values will be replaced with 0.0.
+    posinf (float or optional): Value to be used to fill positive infinity values.
```
posinf (Number, optional): if a Number, the value to replace positive infinity values with. If None, positive infinity is replaced with the greatest finite value representable by :attr:\`input\`'s dtype. Default is None.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,45 @@ def merge_dicts(*dicts):
+    posinf (float or optional): Value to be used to fill positive infinity values.
+    If no value is passed then positive infinity values will be replaced by the largest
+    finite floating point values representable by x.dtype.
+    neginf (float or optional): Value to be used to fill negative infinity values.
```
See comment on posinf above.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,45 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> x = torch.tensor([float('nan'), float('inf'), -float('inf'), 3.14])
```
Nice example.

mruberry commented on test/test_torch.py:
```diff
@@ -18225,6 +18225,40 @@ def test_nansum(self, device, dtype):
 
         self.compare_with_numpy(torch.nansum, np.nansum, x, device, dtype)
 
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes(include_bfloat16=False)))
```
I like this test but would like to reconsider using the OpInfo pattern (discussion in the conversation).

kshitij12345 replied:
How about keeping this test which verifies it with arguments and adding entry for `nan_to_num`  to existing `OpInfo` tests.

mruberry replied:
Sounds good.

mruberry commented:
This PR looks really good. I asked a few questions and made some suggestions.

As for testing, I'd like to add an OpInfo for this function to test the default behavior, then a more specialized test for the optional parameters, like the one you've written already, can be added to test_unary_ufuncs.py. The existing unary ufunc tests will cover a lot of the behavior that's being tested already, so I think this PR can add a simpler test without worrying about out= handling or tensors with strange shapes. 

What are your thoughts on that approach? 

The OpInfo for nan_to_num can also override `sample_inputs()` to return samples that use the optional arguments.

kshitij12345 committed:
address comment

* use C10_HOST_DEVICE.
* use static_cast.

kshitij12345 committed:
address comment

* update docs

kshitij12345 committed:
update UnaryUfuncInfo DB

kshitij12345 committed:
Merge branch 'master' into develop/numpy/nan_to_num

kshitij12345 committed:
update types for nan_to_num

kshitij12345 committed:
try fixing docs warnings

kshitij12345 committed:
try fixing doc warnings

kshitij12345 committed:
try fixing docs - esacpe `s`

kshitij12345 commented:
@mruberry PTAL :)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -373,6 +373,34 @@ Tensor& logit_(Tensor& self, c10::optional<double> eps) {
+    c10::optional<double> nan,
+    c10::optional<double> pos_inf,
+    c10::optional<double> neg_inf) {
+  auto iter = TensorIterator::unary_op(result, self);
```
cc @ezyang @anjali411 this is actually another case where we could consider returning self (for integer tensors)

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -371,6 +371,37 @@ static void polygamma_kernel(TensorIterator& iter, int64_t n) {
+    c10::optional<double> pos_inf,
+    c10::optional<double> neg_inf) {
+  if (c10::isIntegralType(iter.dtype(), true)) {
+    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "nan_to_num_cuda", [&]() {
```
"nan_to_num_cuda" -> "nan_to_num"

This is the CPU version.

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -371,6 +371,37 @@ static void polygamma_kernel(TensorIterator& iter, int64_t n) {
+          : std::numeric_limits<scalar_t>::lowest();
+
+      cpu_kernel(iter, [=](scalar_t a) -> scalar_t {
+        return at::_isfinite(a)
```
This check makes conceptual sense, although I wonder if isfinite is performing all the below checks, anyway, so gating those checks on it is redundant. Maybe the compiler is clever enough to store the result, though.

kshitij12345 replied:
Yeah. I was wondering the same. But this way seemed much easier to grasp.

kshitij12345 replied:
You are indeed correct that this wasn't the best way to go about it (as per the benchmark). Have updated the kernel implementation now. Thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -18229,6 +18229,40 @@ def test_nansum(self, device, dtype):
         self.compare_with_numpy(torch.nansum, np.nansum, x, device, dtype)
 
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes(include_bfloat16=False)))
+    def test_nan_to_num(self, device, dtype):
```
Move this test to test_unary_ufuncs.py

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,41 @@ def merge_dicts(*dicts):
+    {input}
+    nan (Number, optional): the value to replace :literal:`NaN`\s with. Default is zero.
+    posinf (Number, optional): if a Number, the value to replace positive infinity values with.
+        If None, positive infinity is replaced with the greatest finite value representable by :attr:`input`'s dtype.
```
"positive infinity" -> "positive infinity values are"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,41 @@ def merge_dicts(*dicts):
+           r"""
+nan_to_num(input, nan=0.0, posinf=None, neginf=None, *, out=None) -> Tensor
+
+Replaces :literal:`NaN`, positive infinity, and negative infinity values in :attr:input
```
:attr:input -> :attr:\`input\`

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,41 @@ def merge_dicts(*dicts):
+
+Args:
+    {input}
+    nan (Number, optional): the value to replace :literal:`NaN`\s with. Default is zero.
```
is the "\s" necessary to render it properly?

kshitij12345 replied:
Yeah. Without that CI failing due to warnings. 
Will post the reference stackoverflow link explaining the same if I find it.

kshitij12345 replied:
https://stackoverflow.com/questions/58633196/restructuredtext-inline-code-followed-by-s-results-in-inline-interpreted-text

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5278,6 +5278,41 @@ def merge_dicts(*dicts):
+        If None, positive infinity is replaced with the greatest finite value representable by :attr:`input`'s dtype.
+        Default is None.
+    neginf (Number, optional): if a Number, the value to replace negative infinity values with.
+        If None, negative infinity is replaced with the lowest finite value representable by :attr:`input`'s dtype.
```
"negative infinity" -> "negative infinity values are"

kshitij12345 replied:
Done.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2634,6 +2634,13 @@ bool any_variable_defined(variable_list& variables) {
 }
 
+at::Tensor nan_to_num_backward(const at::Tensor& grad, const at::Tensor& self) {
+  auto nan_mask = at::isnan(self);
```
Instead of isnan and isinf, can you use https://pytorch.org/docs/master/generated/torch.isfinite.html?highlight=isfinite#torch.isfinite ?

kshitij12345 replied:
Ah. Totally! Embarrassed to have missed that!

mruberry commented:
Hey @kshitij12345, overall continues to look great, of course. Just a few questions/suggestions.

kshitij12345 committed:
update gradient formula

kshitij12345 committed:
fix kernel name

kshitij12345 committed:
update docs

kshitij12345 committed:
replace dispatch for intergers with inplace copy_

kshitij12345 committed:
kMerge branch 'develop/numpy/nan_to_num' into develop/numpy/nan_to_num

kshitij12345 committed:
move test to test_unary_funcs

kshitij12345 committed:
Merge branch 'master' into develop/numpy/nan_to_num

kshitij12345 commented:
**On CPU**

With 
```c++
    cpu_kernel(iter, [=](scalar_t a) -> scalar_t {
      return (
          at::_isnan(a)
              ? nan_replacement
              : (a == std::numeric_limits<scalar_t>::infinity()
                     ? pos_inf_replacement
                     : (a == -std::numeric_limits<scalar_t>::infinity()
                            ? neg_inf_replacement
                            : a)));
```

```
250 / 250
Average attemts per valid config: 24.1
 Best:                   Size    Shape             X order        Y order
--------------------------------------------------------------------------------
 4.6 ns / element       88236 |   57, 1548       | contiguous   | 
 4.6 ns / element       85680 |  272,  315       | [1 0]        | 
 4.6 ns / element      114070 | 3355,   34       | contiguous   | 
 4.6 ns / element      109872 |  763,  144       | contiguous   | 
 4.6 ns / element      129925 | 5197,   25       | contiguous   | 
 4.6 ns / element      114528 | 3579,   32       | contiguous   | 
 4.6 ns / element      103488 |   56,   88,   21 | contiguous   | 
 4.6 ns / element       97682 |  578,  169       | contiguous   | 
 4.6 ns / element      127932 |   21, 6092       | contiguous   | 
 4.6 ns / element       91840 |  328,  280       | contiguous   | 

Worst:                   Size    Shape             X order        Y order
--------------------------------------------------------------------------------
 4.6 ns / element       80325 |   25,  189,   17 | contiguous   | 
 4.6 ns / element       78645 |   49, 1605       | contiguous   | 
 4.6 ns / element       93024 | 1224,   76       | contiguous   | 
 4.6 ns / element       69608 |   88,  791       | contiguous   | 
 4.6 ns / element       83204 |   62, 1342       | [1 0]        | 
 4.7 ns / element       78336 |   72, 1088       | contiguous   | 
 4.7 ns / element      118020 | 5901,   20       | contiguous   | 
 4.7 ns / element      122960 |  106,   40,   29 | contiguous   | 
 4.7 ns / element       66848 | 4178,   16       | contiguous   | 
 4.9 ns / element       69108 | 1772,   39       | contiguous   | 
```

With
```c++
cpu_kernel(iter, [=](scalar_t a) -> scalar_t {
      return at::_isfinite(a)
          ? a
          : (at::_isnan(a) ? nan_replacement
                           : (a == std::numeric_limits<scalar_t>::infinity()
                                  ? pos_inf_replacement
                                  : neg_inf_replacement));
```

```
250 / 250
Average attemts per valid config: 24.1
 Best:                   Size    Shape             X order        Y order
--------------------------------------------------------------------------------
 6.3 ns / element       66066 |  462,  143       | contiguous   | 
 6.3 ns / element      103488 |   56,   88,   21 | contiguous   | 
 6.3 ns / element       69930 |  945,   74       | contiguous   | 
 6.3 ns / element       74316 |   44, 1689       | contiguous   | 
 6.3 ns / element      114528 | 3579,   32       | contiguous   | 
 6.3 ns / element       88236 |   57, 1548       | contiguous   | 
 6.3 ns / element      109872 |  763,  144       | contiguous   | 
 6.3 ns / element       66880 |  190,  352       | contiguous   | 
 6.3 ns / element      114730 |   35, 3278       | contiguous   | 
 6.3 ns / element       76342 |   41, 1862       | contiguous   | 

Worst:                   Size    Shape             X order        Y order
--------------------------------------------------------------------------------
 6.4 ns / element       78336 |   72, 1088       | contiguous   | 
 6.4 ns / element      115080 | 5754,   20       | contiguous   | 
 6.4 ns / element       80080 | 5005,   16       | contiguous   | 
 6.5 ns / element       79968 |  392,  204       | contiguous   | 
 6.5 ns / element       97482 |  422,  231       | contiguous   | 
 6.5 ns / element      127932 |   21, 6092       | contiguous   | 
 8.0 ns / element      104348 | 1373,   76       | [1 0]        | 
 8.1 ns / element       80166 |  186,  431       | [1 0]        | 
 8.3 ns / element       89870 |  430,  209       | contiguous   | 
 8.3 ns / element      102752 |  247,   16,   26 | contiguous   | 
```

[benchmark.txt](https://github.com/pytorch/pytorch/files/5286309/benchmark.txt)

kshitij12345 committed:
update kernel impl

kshitij12345 committed:
remove unused _isfinite

kshitij12345 commented:
@mruberry PTAL :)

Have addressed the comments as well as updated the kernel.

Errors are not related to this PR. Will rebase after addressing new comments if any. 

Thanks!

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -377,6 +378,36 @@ def test_batch_vs_slicing(self, device, dtype, op):
+            for x in generate_numeric_tensors(dtype=dtype, device=device):
+                if not contiguous:
+                    x = x.T
+                self.compare_with_numpy(torch.nan_to_num, np.nan_to_num, x)
```
Will this and the next comparison be covered by the existing unary ufunc tests?

That is, I'm wondering if only the tests where the extra args are set are needed in this test?

kshitij12345 replied:
I am not sure about method variant but `torch.nan_to_num` is verified. You know where the

mruberry replied:
Variants are tested for consistency here: https://github.com/pytorch/pytorch/blob/7cde662f08b226fb8d7f8e99dbd6f40dbe2dca16/test/test_unary_ufuncs.py#L200

Which is not the most thorough test, but we didn't want to run the tests again for method variants because it would greatly increase test time (and cost).

kshitij12345 replied:
Ah. Will remove those two then. Thanks!

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -377,6 +378,36 @@ def test_batch_vs_slicing(self, device, dtype, op):
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes(include_bfloat16=False)))
+    def test_nan_to_num(self, device, dtype):
+        for contiguous in [False, True]:
+            for x in generate_numeric_tensors(dtype=dtype, device=device):
```
I wonder if this test should use a different generator with more non-finite values, since the numeric tensors will be tested elsewhere?

kshitij12345 replied:
I think this should be fine as the generator has tensors with extremals, otherwise we'll will be creating almost similar function which does the same thing (takes dtype and device) and generates tensor. Let me know what you feel about that.

mruberry replied:
I think you can just call make_tensor:

https://github.com/pytorch/pytorch/blob/7cde662f08b226fb8d7f8e99dbd6f40dbe2dca16/torch/testing/_internal/common_utils.py#L1432

And add the extremal values if it's float. You could create a smaller 64x64 tensor, for example. generate_numeric_tensors will create a very large tensor.

kshitij12345 replied:
Oh. Sure. Thanks!

mruberry commented:
Thanks for the benchmark. PR continues to look good, of course. I just have a couple questions about refining the test in test_unary_ufunc.py to ensure it's not redundant with the other OpInfo tests. Maybe just testing some smaller tensors with more non-finite values would be best?

kshitij12345 commented:
Gentle Ping :)

kshitij12345 committed:
address comment

kshitij12345 committed:
Merge branch 'master' into develop/numpy/nan_to_num

mruberry reviewed (approved):
Awesome! Exciting to have torch.nan_to_num!

kshitij12345 commented:
Gentle Ping :)

mruberry commented:
> Gentle Ping :)

Thanks for the ping, and sorry for the delay. It's been a very busy week with the branch cut. Things should get back to normal soon.

mruberry commented:
Congrats @kshitij12345! This is one of, if not the first, new function to be added to PyTorch 1.8!
